### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by nasgull**

### CODE REVIEW

File: Example/index.php
Comment (Bukashk0zzz) [inline Example/index.php:1]: Let's make folder with small letters.

File: src/EventData.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    private string $pagePath;
+    private string $pageReferrer;
+    private string $pageTitle;
+    private string $screenResolution;
+    private string $userAgent;
+    private array $userData;
+    private string $userId;
+    private string $viewportSize;
+
+    public function __construct(array $data)
```
Comment (Bukashk0zzz) [inline src/EventData.php]: Need possibility to send custom data.
now only predefined will be send.

File: src/StapeSGTM.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        );
+
+        $this->client = new Client();
+        $this->validateConfig();
+    }
+
+	/**
+	 * @throws SGTMException
+	 */
+	public function sendEventData(string $eventName, array $eventData = []): bool
```
Comment (Bukashk0zzz) [inline src/StapeSGTM.php]: Add ability to send request async.

Comment (Bukashk0zzz) [reply src/StapeSGTM.php]: Add callback functionality

File: tests/UserDataImpl.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
```
Comment (Bukashk0zzz) [inline tests/UserDataImpl.php]: Add tests to all classes.

File: .gitignore
Comment (Bukashk0zzz) [inline .gitignore:1]: Add .github pipeline that run tests

File: src/StapeSGTM.php (hunk 2 above)
Comment (kHorozhanov) [inline src/StapeSGTM.php]: is it possible to replace `array $eventData` with something like `Partial<EventData> $eventData`?

File: src/StapeSGTM.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Stape\Sgtm;
+
+use GuzzleHttp\Client;
```
Comment (kHorozhanov) [inline src/StapeSGTM.php]: Would it make sense to use Symfony http client instead?

File: src/EventData.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Stape\Sgtm;
+
+class EventData
```
Comment (kHorozhanov) [inline src/EventData.php]: Unnecessary class.
Should we remove it or convert it to an interface that describes known properties and allows the addition of custom properties as well?

File: src/StapeSGTM.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace Stape\Sgtm;
+
+use Symfony\Component\HttpClient\HttpClient;
+
+class StapeSGTM
```
Comment (kHorozhanov) [inline src/StapeSGTM.php:7]: Should we remove this class and rename StapeSGTMClient to StapeSGTM?

It provides an instance of HttpClient, which can be great for dependency inversion, but StapeSGTMClient still uses Symfony HttpClientInterface, so it should be okay to create a HttpClient inside.

File: src/StapeSGTMClient.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+        return $this->sendEvent($url, $data);
+    }
+
+    private function formatUrl(string $eventName, EventData $eventData): array
+    {
+		$params = [];
+        $url = $this->config['gtm_server_domain'] . $this->config['request_path'];
+        $queryString = \parse_url($url, PHP_URL_QUERY);
+		if ($queryString) {
```
Comment (kHorozhanov) [inline src/StapeSGTMClient.php]: Please add some code-style checks

File: src/StapeSGTMClient.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+		}
+
+        $params['v'] = $this->config['protocol_version'];
+        $params['event_name'] = \urlencode($eventName);
+
+        if ($this->config['richsstsse']) {
+            $params['richsstsse'] = '';
+        }
+
+        $url = $url . "&" . \http_build_query($params);
```
Comment (kHorozhanov) [inline src/StapeSGTMClient.php]: should work if the```url``` (request_path) doesn't contains query params yet

File: src/StapeSGTMClient.php [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+        $postData = [
+            'event_name' => $eventName,
```
Comment (kHorozhanov) [inline src/StapeSGTMClient.php:72]: add ```v```(protocol_version) property to ```$postData```

File: src/StapeSGTMClient.php [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        ];
+
+        $postData = ['body' => \array_merge($postData, $eventData->getData())];
```
Comment (kHorozhanov) [inline src/StapeSGTMClient.php:76]: ```event_name``` from ```$postData``` should override ```event_name``` in ```$eventData```

File: src/EventData.php (hunk 5 above)
Comment (kHorozhanov) [reply src/EventData.php]: @Bukashk0zzz @nasgull 
Or should we revert to the previous version of EventData and provide an example of how to extend it?

File: tests/StapeSGTMTest.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+			'client_id' => '123456',
+			'currency' => 'USD',
+			'ip_override' => '79.144.123.69',
+			'language' => 'en',
+			'page_encoding' => 'UTF-8',
+			'page_hostname' => 'Stape',
+			'page_location' => 'http://stape.io',
+		]);
+
+		$this->assertEquals(200, $response->getStatusCode());
```
Comment (kHorozhanov) [inline tests/StapeSGTMTest.php]: Add request path check

File: src/EventData.php (hunk 5 above)
Comment (nastia-vinnytska) [reply src/EventData.php]: add EventDataInterface and UserDataInterface with default properties + EventData that implements EventDataInterface?

File: src/StapeSGTMClient.php [hunk 12]
```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Stape\Sgtm;
+
+
+use Composer\Util\Http\CurlResponse;
```
Comment (nastia-vinnytska) [inline src/StapeSGTMClient.php]: remove as unused

File: src/StapeSGTMClient.php [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        try {
+            return $this->client->request('POST', $url, $data);
+        } catch (\Throwable $e) {
+            $response = $e->getResponse();
+            $status = $response ? $response->getStatusCode() : null;
+            $statusText = $response ? $response->getReasonPhrase() : null;
+
+            $error = $response ? (string) $response->getBody() : null;
+
+            throw new SGTMException($status, $statusText, $error, "HTTP request failed {$status}");
```
Comment (nastia-vinnytska) [inline src/StapeSGTMClient.php]: http exception will not be cathced while async requests :(

File: src/StapeSGTMClient.php [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+	 * @throws SGTMException
+	 */
+	private function sendEvent(string $url, array $data): ResponseInterface
+    {
+        try {
+            return $this->client->request('POST', $url, $data);
+        } catch (\Throwable $e) {
+            $response = $e->getResponse();
+            $status = $response ? $response->getStatusCode() : null;
+            $statusText = $response ? $response->getReasonPhrase() : null;
```
Comment (nastia-vinnytska) [inline src/StapeSGTMClient.php]: have you checked if getReasonPhrase() is an existing method?

File: src/StapeSGTMClient.php [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+	 * @throws SGTMException
+	 */
+	public function validateConfig(): void
+    {
+        if (
+            !isset($this->config['gtm_server_domain']) ||
+            !preg_match('/^https:\/\//', $this->config['gtm_server_domain']) ||
+            preg_match('/^\//', $this->config['gtm_server_domain'])
+        ) {
+            throw new SGTMException(null, null, null, 'You did not fill in the variable gtm_server_domain. Example: https://gtm.stape.io');
```
Comment (nastia-vinnytska) [inline src/StapeSGTMClient.php]: let's make these params optional for SGTMException
```
private ?int $status;
private ?string $statusText;
private ?string $error;
```

File: tests/StapeSGTMTest.php [hunk 16]
```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Stape\tests;
```
Comment (nastia-vinnytska) [inline tests/StapeSGTMTest.php]: *Tests

File: tests/StapeSGTMTest.php [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+			'language' => 'en',
+			'page_encoding' => 'UTF-8',
+			'page_hostname' => 'Stape',
+			'page_location' => 'http://stape.io',
+		]);
+
+		$this->assertEquals(200, $response->getStatusCode());
+	}
+
+	public function testBase64(): void
```
Comment (nastia-vinnytska) [inline tests/StapeSGTMTest.php]: mb, move transform tests to separate test class?

**Pull request merged**